elements/lisk-tree/src/sparse_merkle_tree/utils.ts

```diff
@@ -13,9 +13,180 @@
  */
 
 import { hash } from '@liskhq/lisk-cryptography';
-import { LEAF_HASH_PREFIX, BRANCH_HASH_PREFIX } from './constants';
+import { objects } from '@liskhq/lisk-utils';
+import { BRANCH_HASH_PREFIX, EMPTY_HASH, LEAF_HASH_PREFIX } from './constants';
+import { InclusionProofQuery, InclusionProofQueryWithHash } from './types';
+
+// Sort queries by the longest binaryBitmap, breaking ties by smaller key.
+// https://github.com/LiskHQ/lips-staging/blob/master/proposals/lip-0039.md#proof-construction
+export const sortByBitmapAndKey = <T extends InclusionProofQuery>(queries: T[]): T[] =>
+	queries.sort((q1, q2) => {
+		const q1BinaryBitmap = bufferToBinaryString(q1.bitmap);
+		const q2BinaryBitmap = bufferToBinaryString(q2.bitmap);
+
+		if (q1BinaryBitmap.length === q2BinaryBitmap.length) {
+			return q1.key.byteLength - q2.key.byteLength;
+		}
+
+		return q2BinaryBitmap.length - q1BinaryBitmap.length;
+	});
+
+// Remove queries that have merged together, keep only those with a different key prefix
+// https://github.com/LiskHQ/lips-staging/blob/master/proposals/lip-0039.md#proof-verification
+export const filterQueries = <T extends InclusionProofQuery>(
+	queries: T[],
+	keyLength: number,
+): T[] => {
+	const uniqueKeys: string[] = [];
+
+	return queries.filter(q => {
+		const binaryBitmap = bufferToBinaryString(q.bitmap);
+		const h = binaryBitmap.length;
+		const binaryKey = binaryExpansion(q.key, keyLength);
+		const keyPrefix = binaryKey.substring(0, h);
+
+		if (!uniqueKeys.includes(keyPrefix)) {
+			uniqueKeys.push(keyPrefix);
+			return true;
+		}
+
+		return false;
+	});
+};
+
+// Checks whether two queries correspond to nodes that are
+// children of the same branch node, with q1 and q2 the left and right
+// child respectively
+// https://github.com/LiskHQ/lips-staging/blob/master/proposals/lip-0039.md#proof-verification
+export const areSiblingQueries = (
+	q1: InclusionProofQuery,
+	q2: InclusionProofQuery,
+	keyLength: number,
+): boolean => {
+	const q1BinaryBitmap = bufferToBinaryString(q1.bitmap);
+	const q2BinaryBitmap = bufferToBinaryString(q2.bitmap);
+
+	if (q1BinaryBitmap.length !== q2BinaryBitmap.length) {
+		return false;
+	}
+
+	const h = q1BinaryBitmap.length;
+	const binaryKey1 = binaryExpansion(q1.key, keyLength);
+	const binaryKey2 = binaryExpansion(q2.key, keyLength);
+
+	// end of string is exclusive
+	const keyPrefix1 = binaryKey1.substring(0, h - 1);
+	const keyPrefix2 = binaryKey2.substring(0, h - 1);
+
+	if (keyPrefix1 !== keyPrefix2) {
+		return false;
+	}
+
+	const d1 = binaryKey1[h];
+	const d2 = binaryKey2[h];
+
+	return (d1 === '0' && d2 === '1') || (d1 === '1' && d2 === '0');
```

Comment:
in the LIP, it says `return d1==0 and d2==1` because q1 and q2 are left and right respectively?

Reply:
Was discussing with @ricott1 and he suggested that we should fix this and make it commutative.